### ISSUE TITLE
Cannot assign client roles to a user when a realm contains more than ~4000 clients

### ISSUE THREAD

**Issue by reda-alaoui** [keycloak/keycloak#23404]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

storage

### Describe the bug

One of our realms contains 8754 clients.

On this realm, we are unable to assign a client role to a user.

### Version

22.0.1

### Expected behavior

We should be able to assign client roles to a user even when the database contains many clients.

### Actual behavior

In order to assign a client role to a user, we have to list them. On a realm with many clients, the client selection screen freezes and never loads:

![image](https://github.com/keycloak/keycloak/assets/2890843/92fede83-ebb7-400e-96b5-c4db23cef2bf)

The hanging http request is `GET https://keycloak.acme.com/auth/admin/realms/general/ui-ext/available-roles/users/363b9993-ee55-4aff-a415-439ea28396f6?first=0&max=101&search=`

After 5.3 minutes, the http request fails with an internal 500 error and the UI displays an error:

![image](https://github.com/keycloak/keycloak/assets/2890843/d30f676c-e9fc-42d2-a1b1-df694f717798)

Multiple jstack snapshots always show the same stack:
```java
"executor-thread-109" - Thread t@19291
   java.lang.Thread.State: RUNNABLE
        at org.hibernate.persister.entity.AbstractEntityPersister.getPropertyValue(AbstractEntityPersister.java:4227)
        at org.hibernate.persister.entity.EntityPersister.getValue(EntityPersister.java:920)
        at org.hibernate.engine.internal.Cascade.cascade(Cascade.java:184)
        at org.hibernate.event.internal.AbstractSaveEventListener.cascadeAfterSave(AbstractSaveEventListener.java:490)
        at org.hibernate.event.internal.DefaultPersistEventListener.justCascade(DefaultPersistEventListener.java:161)
        at org.hibernate.event.internal.DefaultPersistEventListener.entityIsPersistent(DefaultPersistEventListener.java:154)
        at org.hibernate.event.internal.DefaultPersistEventListener.persist(DefaultPersistEventListener.java:90)
        at org.hibernate.event.internal.DefaultPersistEventListener.onPersist(DefaultPersistEventListener.java:77)
        at org.hibernate.internal.SessionImpl$$Lambda$1923/0x00007f687ce38230.applyEventToListener(Unknown Source)
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:138)
        at org.hibernate.internal.SessionImpl.persistOnFlush(SessionImpl.java:819)
        at org.hibernate.engine.spi.CascadingActions$8.cascade(CascadingActions.java:333)
        at org.hibernate.engine.spi.CascadingActions$8.cascade(CascadingActions.java:323)
        at org.hibernate.engine.internal.Cascade.cascadeToOne(Cascade.java:513)
        at org.hibernate.engine.internal.Cascade.cascadeAssociation(Cascade.java:434)
        at org.hibernate.engine.internal.Cascade.cascadeProperty(Cascade.java:220)
        at org.hibernate.engine.internal.Cascade.cascadeComponent(Cascade.java:405)
        at org.hibernate.engine.internal.Cascade.cascadeProperty(Cascade.java:241)
        at org.hibernate.engine.internal.Cascade.cascade(Cascade.java:153)
        at org.hibernate.event.internal.AbstractFlushingEventListener.cascadeOnFlush(AbstractFlushingEventListener.java:155)
        at org.hibernate.event.internal.AbstractFlushingEventListener.prepareEntityFlushes(AbstractFlushingEventListener.java:145)
        at org.hibernate.event.internal.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:79)
        at org.hibernate.event.internal.DefaultAutoFlushEventListener.onAutoFlush(DefaultAutoFlushEventListener.java:48)
        at org.hibernate.internal.SessionImpl$$Lambda$1831/0x00007f687cdb9600.accept(Unknown Source)
        at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
        at org.hibernate.internal.SessionImpl.autoFlushIfRequired(SessionImpl.java:1375)
        at org.hibernate.sql.exec.internal.JdbcSelectExecutorStandardImpl.scroll(JdbcSelectExecutorStandardImpl.java:115)
        at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.lambda$new$1(ConcreteSqmSelectQueryPlan.java:136)
        at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan$$Lambda$1816/0x00007f687cdae3d8.interpret(Unknown Source)
        at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.withCacheableSqmInterpretation(ConcreteSqmSelectQueryPlan.java:303)
        at org.hibernate.query.sqm.internal.ConcreteSqmSelectQueryPlan.performScroll(ConcreteSqmSelectQueryPlan.java:252)
        at org.hibernate.query.sqm.internal.QuerySqmImpl.doScroll(QuerySqmImpl.java:629)
        at org.hibernate.query.spi.AbstractSelectionQuery.scroll(AbstractSelectionQuery.java:445)
        at org.hibernate.query.spi.AbstractSelectionQuery.stream(AbstractSelectionQuery.java:457)
        at org.hibernate.query.Query.getResultStream(Query.java:168)
        at org.keycloak.models.jpa.JpaRealmProvider.getClientScopes(JpaRealmProvider.java:986)
        at org.keycloak.storage.ClientStorageManager.getClientScopes(ClientStorageManager.java:212)
        at org.keycloak.models.cache.infinispan.RealmCacheSession.getClientScopes(RealmCacheSession.java:1358)
        at org.keycloak.models.jpa.ClientAdapter.getClientScopes(ClientAdapter.java:366)
        at org.keycloak.models.cache.infinispan.entities.CachedClient.<init>(CachedClient.java:112)
        at org.keycloak.models.cache.infinispan.RealmCacheSession.cacheClient(RealmCacheSession.java:1152)
        at org.keycloak.models.cache.infinispan.RealmCacheSession.getClientById(RealmCacheSession.java:1114)
        at org.keycloak.models.delegate.ClientModelLazyDelegate$WithId.lambda$new$0(ClientModelLazyDelegate.java:53)
        at org.keycloak.models.delegate.ClientModelLazyDelegate$WithId$$Lambda$2700/0x00007f687d16fdd8.get(Unknown Source)
        at org.keycloak.models.delegate.ClientModelLazyDelegate.getDelegate(ClientModelLazyDelegate.java:83)
        at org.keycloak.models.delegate.ClientModelLazyDelegate.getRolesStream(ClientModelLazyDelegate.java:579)
        at org.keycloak.admin.ui.rest.RoleMappingResource$$Lambda$3072/0x00007f687d272b70.apply(Unknown Source)
        at java.base@17.0.8/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273)
        at java.base@17.0.8/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base@17.0.8/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1856)
        at java.base@17.0.8/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:129)
        at java.base@17.0.8/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
        at java.base@17.0.8/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
        at java.base@17.0.8/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base@17.0.8/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
        at java.base@17.0.8/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base@17.0.8/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
        at org.keycloak.utils.ClosingStream.collect(ClosingStream.java:182)
        at org.keycloak.admin.ui.rest.RoleMappingResource.mapping(RoleMappingResource.java:36)
        at org.keycloak.admin.ui.rest.AvailableRoleMappingResource.listCompositeUserRoleMappings(AvailableRoleMappingResource.java:158)
        at java.base@17.0.8/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base@17.0.8/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base@17.0.8/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base@17.0.8/java.lang.reflect.Method.invoke(Method.java:568)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:154)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:118)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:560)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:452)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:413)
        at org.jboss.resteasy.core.ResourceMethodInvoker$$Lambda$2361/0x00007f687cfbdcb8.get(Unknown Source)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
        - locked <75c69112> (a org.jboss.resteasy.core.interception.jaxrs.PostMatchContainerRequestContext)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:415)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:378)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:174)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:142)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:168)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:142)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:168)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:142)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:168)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:131)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:33)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:429)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:240)
        at org.jboss.resteasy.core.SynchronousDispatcher$$Lambda$2359/0x00007f687cfbbf88.run(Unknown Source)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:154)
        at org.jboss.resteasy.core.SynchronousDispatcher$$Lambda$2360/0x00007f687cfbc898.get(Unknown Source)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:321)
        - locked <3d016fe> (a org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:157)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:229)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:82)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:147)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:84)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:44)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
        at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
        at io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1.handle(HttpServerCommonHandlers.java:58)
        at io.quarkus.vertx.http.runtime.options.HttpServerCommonHandlers$1.handle(HttpServerCommonHandlers.java:36)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:177)
        at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
        at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$0(QuarkusRequestFilter.java:82)
        at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter$$Lambda$2355/0x00007f687cfb1b20.run(Unknown Source)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base@17.0.8/java.lang.Thread.run(Thread.java:833)
```

### How to Reproduce?

Follow https://github.com/Cosium/keycloak-issue-21010 , then try to assign a client role to a user.

### Anything else?

This is the same kind of issue as https://github.com/keycloak/keycloak/issues/21010

**Comment by stianst** [comment]
@ssilvert @ahus1 this is probably something we should address in Keycloak 22. @sschu is saying he could work on this if we can have some conversations on how to proceed.

We may want to add something around this in the Admin APIs, instead of in the Admin UI Ext APIs.

**Comment by ssilvert** [comment]
@ahus1 @sschu we would love to see something added to the Admin APIs for this as @stianst suggests.  

Please work with @edewit when you are ready to proceed with UI changes.

**Comment by reda-alaoui** [comment]
If anyone is interested, here is my workaround.

The following goes into a plugin:

```java
package com.acme.keycloak_plugin.issue_23404;

import static org.keycloak.utils.StreamsUtil.closing;

import jakarta.persistence.EntityManager;
import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.CriteriaQuery;
import jakarta.persistence.criteria.Root;
import jakarta.persistence.criteria.Subquery;
import jakarta.ws.rs.ForbiddenException;
import jakarta.ws.rs.NotFoundException;
import java.util.List;
import java.util.Objects;
import org.keycloak.admin.ui.rest.model.ClientRole;
import org.keycloak.admin.ui.rest.model.RoleMapper;
import org.keycloak.connections.jpa.JpaConnectionProvider;
import org.keycloak.models.KeycloakSession;
import org.keycloak.models.RealmModel;
import org.keycloak.models.UserModel;
import org.keycloak.models.UserProvider;
import org.keycloak.models.jpa.RoleAdapter;
import org.keycloak.models.jpa.entities.RoleEntity;
import org.keycloak.models.jpa.entities.UserRoleMappingEntity;
import org.keycloak.services.resources.admin.permissions.AdminPermissionEvaluator;

/**
 * A performant alternative to {@link org.keycloak.admin.ui.rest.AvailableRoleMappingResource} to
 * workaround <a href="https://github.com/keycloak/keycloak/issues/23404">Cannot assign client roles
 * to a user when a realm contains more than ~4000 clients</a>
 *
 * @author Réda Housni Alaoui
 */
// TODO Remove once [Cannot assign client roles to a user when a realm contains more than ~4000
//  clients](https://github.com/keycloak/keycloak/issues/23404) is fixed.
public class PerformantAvailableRoleMappingResource {

  private final KeycloakSession session;
  private final RealmModel realm;
  private final AdminPermissionEvaluator auth;

  public PerformantAvailableRoleMappingResource(
      KeycloakSession session, RealmModel realm, AdminPermissionEvaluator auth) {
    this.session = session;
    this.realm = realm;
    this.auth = auth;
  }

  /**
   * @see
   *     org.keycloak.admin.ui.rest.AvailableRoleMappingResource#listCompositeUserRoleMappings(String,
   *     long, long, String)
   */
  public final List<ClientRole> listCompositeUserRoleMappings(
      String id, long first, long max, String search) {
    UserProvider users = Objects.requireNonNull(session).users();
    UserModel userModel = users.getUserById(this.realm, id);
    if (userModel == null) {
      if (auth.users().canQuery()) {
        throw new NotFoundException("User not found");
      } else {
        throw new ForbiddenException();
      }
    }

    this.auth.users().requireView(userModel);
    return this.doListCompositeUserRoleMappings(userModel, first, max, search);
  }

  private List<ClientRole> doListCompositeUserRoleMappings(
      UserModel userModel, long first, long max, final String search) {

    EntityManager entityManager =
        session.getProvider(JpaConnectionProvider.class).getEntityManager();
    CriteriaBuilder cb = entityManager.getCriteriaBuilder();
    CriteriaQuery<RoleEntity> query = cb.createQuery(RoleEntity.class);
    Root<RoleEntity> root = query.from(RoleEntity.class);

    Subquery<UserRoleMappingEntity> subquery = query.subquery(UserRoleMappingEntity.class);
    Root<UserRoleMappingEntity> subqueryRoot = subquery.from(UserRoleMappingEntity.class);

    subquery.where(
        cb.equal(root.get("id"), subqueryRoot.get("roleId")),
        cb.equal(subqueryRoot.get("user").get("id"), userModel.getId()));

    query
        .where(
            cb.equal(root.get("realmId"), realm.getId()),
            cb.isTrue(root.get("clientRole")),
            cb.or(
                cb.like(root.get("clientId"), "%" + search + "%"),
                cb.like(root.get("name"), "%" + search + "%")),
            cb.exists(subquery).not())
        .orderBy(cb.asc(root.get("clientId")), cb.asc(root.get("name")));

    return closing(entityManager.createQuery(query).getResultStream())
        .map(roleEntity -> new RoleAdapter(session, realm, entityManager, roleEntity))
        .filter(auth.roles()::canMapRole)
        .map(roleModel -> RoleMapper.convertToModel(roleModel, realm.getClientsStream()))
        .skip(first)
        .limit(max)
        .toList();
  }
}
```

This following goes into a ByteBuddy java agent:

```java
package com.acme.keycloak_agent;

import static net.bytebuddy.matcher.ElementMatchers.isPublic;
import static net.bytebuddy.matcher.ElementMatchers.named;
import static net.bytebuddy.matcher.ElementMatchers.returns;
import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
import static net.bytebuddy.matcher.ElementMatchers.takesArguments;

import java.lang.instrument.Instrumentation;
import java.lang.reflect.InvocationTargetException;
import java.security.ProtectionDomain;
import java.util.List;
import net.bytebuddy.agent.builder.AgentBuilder;
import net.bytebuddy.asm.Advice;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.dynamic.DynamicType;
import net.bytebuddy.implementation.bytecode.assign.Assigner;
import net.bytebuddy.utility.JavaModule;

/**
 * @author Réda Housni Alaoui
 */
// TODO Remove once [Cannot assign client roles to a user when a realm contains more than ~4000
//  clients](https://github.com/keycloak/keycloak/issues/23404) is fixed.
public class AvailableRoleMappingResourceInterceptor {

  private static final String INTERCEPTED_METHOD_NAME = "listCompositeUserRoleMappings";

  private AvailableRoleMappingResourceInterceptor() {}

  public static void registerTo(Instrumentation instrumentation) {
    AgentBuilders.create()
        .type(named("org.keycloak.admin.ui.rest.AvailableRoleMappingResource"))
        .transform(new PerformanceTransformer())
        .installOn(instrumentation);
  }

  private static class PerformanceTransformer implements AgentBuilder.Transformer {

    @Override
    public DynamicType.Builder<?> transform(
        DynamicType.Builder<?> builder,
        TypeDescription typeDescription,
        ClassLoader classLoader,
        JavaModule module,
        ProtectionDomain protectionDomain) {
      return builder
          .method(
              named("listCompositeUserRoleMappings")
                  .and(isPublic())
                  .and(takesArguments(4))
                  .and(returns(List.class))
                  .and(takesArgument(0, String.class))
                  .and(takesArgument(1, Long.TYPE))
                  .and(takesArgument(2, Long.TYPE))
                  .and(takesArgument(3, String.class)))
          .intercept(Advice.to(PerformanceAdvice.class));
    }
  }

  private static class PerformanceAdvice {
    @Advice.OnMethodEnter(skipOn = SkipInstrumentedMethodExecution.class)
    static SkipInstrumentedMethodExecution enter() {
      return new SkipInstrumentedMethodExecution();
    }

    @Advice.OnMethodExit
    static void exit(
        @Advice.FieldValue(value = "session", typing = Assigner.Typing.DYNAMIC) Object session,
        @Advice.FieldValue(value = "realm", typing = Assigner.Typing.DYNAMIC) Object realm,
        @Advice.FieldValue(value = "auth", typing = Assigner.Typing.DYNAMIC) Object auth,
        @Advice.AllArguments Object[] arguments,
        @Advice.Return(readOnly = false, typing = Assigner.Typing.DYNAMIC) Object result) {
      try {
        Class<?> resourceClass =
            Class.forName(
                "com.acme.keycloak_plugin.issue_23404.PerformantAvailableRoleMappingResource");
        Object resource =
            resourceClass
                .getConstructor(
                    Class.forName("org.keycloak.models.KeycloakSession"),
                    Class.forName("org.keycloak.models.RealmModel"),
                    Class.forName(
                        "org.keycloak.services.resources.admin.permissions.AdminPermissionEvaluator"))
                .newInstance(session, realm, auth);

        // This harmless looking assignation swaps the original realm model with the new
        // one.
        // This is dark magic
        result =
            resourceClass
                .getMethod(
                    INTERCEPTED_METHOD_NAME, String.class, Long.TYPE, Long.TYPE, String.class)
                .invoke(resource, arguments);
      } catch (IllegalAccessException
          | NoSuchMethodException
          | InvocationTargetException
          | ClassNotFoundException
          | InstantiationException e) {
        throw new RuntimeException(e);
      }
    }
  }

  public static class SkipInstrumentedMethodExecution {}
}
```

```java
package com.acme.keycloak_agent;

import java.lang.instrument.Instrumentation;

public class PluginAgent {

  private PluginAgent() {}

  public static void premain(String arguments, Instrumentation instrumentation) {
    AvailableRoleMappingResourceInterceptor.registerTo(instrumentation);
  }
}
```

**Comment by ssilvert** [comment]
@ahus1 @sschu @edewit Any progress on this?

**Comment by sschu** [comment]
I am on it. Hopefully I'll have something tomorrow.

**Comment by sschu** [comment]
I have to apologize, I didn't manage to get to this. The new plan is to come up with something on Monday.

**Comment by sschu** [comment]
@mhajas I opened a first draft PR. I would continue the discussion there.

**Comment by reda-alaoui** [comment]
@sschu , it is not exactly that.
Yes, I didn’t change the pagination aspect.
But, the main point is that, instead of issuing one `client role` SQL query per client, I retrieve all client roles of all clients in a *single* SQL query.

TLDR: My workaround fixes the n+1 query issue where `n` is the number of clients.

**Comment by jonkoops** [comment]
Moving this to v23, as we will no longer be doing patch releases for v22.

**Comment by reda-alaoui** [comment]
Please note that the same issue affects client role assignation to groups.

**Comment by sschu** [comment]
@reda-alaoui I updated my PR https://github.com/keycloak/keycloak/pull/24012 and it should also work for groups.
